S3 C++: Replace the remaing c code[WIP] Fixes #11

Thaodan opened the pull request:
Currently so far:
- s3_make_date(): Refactored out

llewelld commented on src/s3access.cpp:
```diff
@@ -44,6 +44,12 @@ S3Access::~S3Access()
     }
 }
 
+s3_date S3Access::now_as_s3date(void) const {
```
For the sake of consistency, could the opening method bracket be put on its own line?

llewelld commented on src/s3access.cpp:
```diff
@@ -110,27 +116,28 @@ QString S3Access::bucket() const
+QNetworkReply *S3Access::performOp(Method method,
+                                   QString const &url, QString const &sign_data,
+                                   s3_date date, QIODevice *in,
+                                   const char *content_md5, const char *content_type)
```
I know there's no offiical style guide or anything, and that it doesn't match the Qt styleguide either, but I'd prefer statements to all be kept on the same line.

Thaodan replied:
I look at the Qt styleguide later however the line was very long making it bad to read.

Thaodan replied:
https://wiki.qt.io/Qt_Coding_Style#Line_breaks

llewelld replied:
Yes, what you've done is correctly in line with the Qt guidelines, but this is one area where the project diverges from them. If you have a look through the rest of the code you should see there's a "no-hard-linebreaks-mid-statement" rule applied.

llewelld commented on src/s3access.cpp:
```diff
@@ -139,8 +146,9 @@ QNetworkReply *S3Access::performOp(Method method, QString const &url, QString co
-    request.setRawHeader("Authorization", QString(QStringLiteral("AWS %1:%2")).arg(m_id).arg(digest).toLocal8Bit());
+    request.setRawHeader("Date", date.toLocal8Bit());
+    request.setRawHeader("Authorization",
+                         QString(QStringLiteral("AWS %1:%2")).arg(m_id).arg(digest).toLocal8Bit());
```
Similarly, if this could be kept on a single line that would be preferable.

llewelld reviewed (changes requested):
This is great, thank you. There are just a couple of style issues to fix.

There's still some lingering C code being called in [`S3ListResult::onFinished()`](https://github.com/llewelld/harbour-contrac/pull/21/files#diff-9b98894de23d3232ab325bb0287743ebR194) that's preventing `s3.h` from being removed. This can be addressed separately, but I guess I wanted to check that this is intentional. Is that correct?

Thaodan committed:
Factor out s3_make_date by refactoring performOp

Moved the creation of the signData string to performOp making
s3_make_date() unneeded. Added signDataKey to allow for optional
signData key for methods like getFile(). Should also make adding
methods that user other Methods easier.

Also cleanes up the converting of date to QString and just converts it
to QBytearray when needed.

Thaodan commented:
I don't have much knowledge of the s3 c code but this could be a replacement for the c code in C++/Qt removing the need for direct openssl use:
https://wiki.qt.io/HMAC-SHA1

llewelld commented:
> I don't have much knowledge of the s3 c code but this could be a replacement for the c code in C++/Qt removing the need for direct openssl use:
> https://wiki.qt.io/HMAC-SHA1

Sorry for taking so long to reply; I hadn't noticed I'd left this question hanging. That's a nice find, but as it happens the openssl linkage will be needed for other things, so I'd prefer to stick with openssl throughout. However the usage in `s3_hmac_sign()` can be achieved with just a single openssl call, so it would be good to simplify it. See for example [its use here](https://github.com/llewelld/harbour-contrac/blob/458fe1d0f5fbebc4e8968779e873f5181ca46630/contracd/src/contrac.cpp#L125).

llewelld commented:
It should be possible to simplify the MD5 code similarly: https://www.openssl.org/docs/man1.1.1/man3/MD5.html

Thaodan committed:
Replace s3_hmac_sign() with HMAC() from OpenSSL

Instead of using s3_hmac_sign() which does everything on its own
resuse HMAC() from OpenSSL. This adds some casting that was done
previously inside s3_hmac_sign() to performOp().
In addition we no longer use the deprecated HMAC_Init(HMAC_CTX *ctx, const void *key, int key_len,
               const EVP_MD *md) (<1.1).

Removes the usage of:
+ s3_hmac_sign()

Contributes to #11

Thaodan commented:
Is the md5 function used somewhere?

llewelld commented:
> Is the md5 function used somewhere?

The only instance I could see was in `s3_put()`, which isn't used yet (and I think probably won't be needed in the future), so feel free to drop both if you also think they're not needed.

Thaodan committed:
Refactor out the usage of the S3 struct

The S3 struct is no longer used since all c-style s3 methods are no
longer used.
Remove all code that related to it.

Removes the usage of:
+ s3_free()
+ s3_init()

Contributes to #11.

llewelld commented on src/s3access.cpp:
```diff
@@ -110,37 +96,60 @@ QString S3Access::bucket() const
+        break;
+    }
+
+    QString signData = methodStr + "\n\n\n" +   + "\n/" + m_bucket + "/";
```
There's a date missing from here. It should be:
```
String signData = methodStr + "\n\n\n" +  date + "\n/" + m_bucket + "/";
```

llewelld commented on src/s3access.cpp:
```diff
@@ -110,37 +96,60 @@ QString S3Access::bucket() const
+        methodStr = "PUT";
+        qDebug() << "PUT request";
+        break;
+    case GET:
```
You can remove the `case GET:` here to match the code further down (or add the same code at line, but best to stay consistent with it).

llewelld commented on src/s3access.cpp:
```diff
@@ -110,37 +96,60 @@ QString S3Access::bucket() const
 {
     QNetworkRequest request;
-    char *digest;
+    unsigned char digest[EVP_MAX_MD_SIZE +1 ]; // +1 for the \0 byte
```
The +1 is no longer needed here (since you're not adding the null byte after all).

llewelld commented on src/s3access.cpp:
```diff
@@ -110,37 +96,60 @@ QString S3Access::bucket() const
+    QString signData = methodStr + "\n\n\n" +   + "\n/" + m_bucket + "/";
+
+    if (signDataKey != nullptr)
+        signData+=signDataKey;
```
For consistency (and because it's a common source of error) I'd be grateful if you could include the curly brackets here.

llewelld commented on src/s3access.cpp:
```diff
@@ -168,7 +177,6 @@ QNetworkReply *S3Access::performOp(Method method, QString const &url, QString co
     }
 
-    free(digest);
 
```
There's an extra line here after you've removed the `free(digest)`. If you could remove it, that would be appreciated.

Thaodan replied:
I don't see that line, I just removed that line here.

llewelld replied:
Yes, I wasn't very clear. I mean if you have this:
```
1.
2. statement
3.
```
and then remove `statement`, you end up with two blank lines:
```
1.
3.
```
but for purely aesthetic reasons there should only be a single blank line after the switch statement.

llewelld reviewed (changes requested):
This is really great and looking nice and clean. I tested it, and with the "date" part added in it worked great too. I made a few style requests, but the only change that's essential to getting it to work is for the "date" line to be fixed.

Thaodan committed:
Add braces to if signDataKey for consistency

llewelld commented:
You still have [WIP] in the title, but this is good enough -- and substantial enough -- to merge in its current form. Did you have other changes in mind? If you did, it might make sense to put them in a follow-up PR, if you prefer?

Thaodan commented:
I'd  replace the s3 buckets struct if you don't mind and merge this 
afterwards.
Then hole code of the s3 folder is no longer needed.
